src/cast_core/src/SamplingLoader.cpp

```diff
@@ -70,42 +70,67 @@ struct SamplingLoader::PhaseConfig {
     PipelineGenerator pipelineSuffixGenerator;
 };
 
+std::vector<bsoncxx::document::value> SamplingLoader::gatherSample(PhaseConfig* config) {
+    mongocxx::pipeline samplePipeline;
+    samplePipeline.sample(config->sampleSize);
+    samplePipeline.project(make_document(kvp("_id", 0)));
+    const auto suffixPipe = pipeline_helpers::makePipeline(config->pipelineSuffixGenerator);
+    samplePipeline.append_stages(suffixPipe.view_array());
+
+    std::vector<bsoncxx::document::value> sampleDocs;
+    const int maxRetries = 3;
+    for (int nRetries = 0; nRetries < maxRetries; ++nRetries) {
+        try {
+            auto cursor =
```

Comment:
If I'm reading the mongocxx code correctly, the cursor's interator is only of type `std::input_iterator_tag` which means the constructor using cursor.begin()/.end() will reallocate the vector O(log N) times (according to https://en.cppreference.com/w/cpp/container/vector/vector). The copy assignment in the line that follows will move from the temporary but it would still waste the allocation that was done when creating sampleDocs. If we want to be more perf-minded here, I'd suggest to reserve sampleDocs's capacity and then push_back into it in a loop from the cursor. Totally your call (for small samples none of this should matter and I'm thinking we should probably add a line to the actor's documentation discouraging usage of large samples, if anything, to reduce the probability of retries).